### DISCUSSION TITLE
First call to `getServerSideProps` is slow

### DISCUSSION THREAD

**Discussion by joaquimds** [vercel/next.js#12447] [Help]
Hi all,

I was just hoping to check if things are working normally or if I've done something wrong. It seems like the first call to `getServerSideProps` takes around 2 seconds when running Next in production mode.

I have two pages, Home and Test, and a `<Link>` from Home to Test. 

Test contains the following code:

    import Link from 'next/link'
    
    const Test = () => (
      <>
        <div className="header">
          <h1>Test</h1>
          <Link href="/">
            <a>Home</a>
          </Link>
        </div>
      </>
    )
    
    export const getServerSideProps = async ({ req, res }) => {
      return { props: {} }
    }
    
    export default Test

Why might `getServerSideProps` take so long to run first time?

Furthermore - the Test page loads almost immediately the second time I navigate to it, so I'm wondering if there's some way to warm whatever cache is providing the speedup?

**Comment by samuelcelko** [comment]
I'm experiencing the same issue. In dev, the first route change takes a while and every subsequent one (even when the server fetched data is different!) is almost instant. 

In prod, every route change to a component with `getServerSideProps` takes the same amount - around 2-3 seconds, even when the data fetched is quite small.

Any ideas on what might be causing this?

**Comment by EnvestorsTeam** [comment]
Also having the same thing, getServerSideProps, seems to just take such a long time?

**Comment by joaquimds** [comment]
I have no fix but I added a loading indicator for UX...

```
  _app.js
  -------
  import NProgress from 'nprogress'
  import 'nprogress/nprogress.css'
  import Router from 'next/router'

  const MyApp = () => {
    useEffect(() => {
      Router.events.on('routeChangeStart', NProgress.start)
      Router.events.on('routeChangeComplete', NProgress.done)
      Router.events.on('routeChangeError', NProgress.done)
      return () => {
        Router.events.off('routeChangeStart', NProgress.start)
        Router.events.off('routeChangeComplete', NProgress.done)
        Router.events.off('routeChangeError', NProgress.done)
      }
    }, [])
  }
```

**Reply by talbertherndon** [reply]
To add to this, just install this package: https://www.npmjs.com/package/nprogress

Then add the code above to you _app.js file, it should work pretty well as an indicator.

**Comment by samuelcelko** [comment]
Same here, I've added an NProgress indicator which makes it a little more obvious that something is being loaded. However, the (bug?) still persist and I'm wondering whether anyone has experienced the same thing and found a solution.

**Reply by simonuvarov** [reply]
Are you using prod in a serverless mode? If so, then it takes some time [to warm up a serverless function](https://www.serverless.com/blog/keep-your-lambdas-warm) (nextjs handler).

**Comment by aminick** [comment]
Same problem, looking for some 'best practices' on this topic

**Comment by tomiiide** [comment]
any fix to this yet?

**Comment by woodybury** [comment]
bump. Sometimes takes > 7 seconds in production even if getServerSideProps doesn't even make an api call!

**Comment by montanaflynn** [comment]
I've noticed this as well

**Comment by sptclr** [comment]
any update for this problems? i have same issue..

**Comment by nastaransamui** [comment]
I have same problem on this with cookies check and Vercel show "504: GATEWAY_TIMEOUT"

**Comment by 2stash** [comment]
Anyone update to latest next.js package? Worked for discussion below.

https://github.com/vercel/next.js/discussions/34540

**Comment by nastaransamui** [comment]
update to "next": "12.1.0" solved my problem.

**Reply by jeffminsungkim** [reply]
It didn't work for me. I'm using v12.3.1 atm.

**Comment by Morbe17** [comment]
Still experiencing this. Could be beacasue the amount of imports?

**Reply by nastaransamui** [reply]
The last part that help me to improve this and not get any error on Vercel was move my dbConnection from getServerSideProps to each API files.

**Comment by jodik** [comment]
Added my other 2 main pages here: https://screenbud.com/s/RVoMPSVDhrc initial loadup time went down from 10-20secs to ~0

**Reply by jodik** [reply]
Maybe that's not even necessary with following next.config.js:
```
   modularizeImports: {
      '@mui/icons-material': {
        transform: '@mui/icons-material/{{member}}',
      },
    },
```

**Comment by mshahzaib101** [comment]
got this solution while searching for this bug

```
const isServerReq = req => !req.url.startsWith('/_next');

function Post({ post }) {
  // Render post...
}

export async function getServerSideProps({ req }) {
  const initData = isServerReq(req) ? await fetch(...) : null;

  return {
    props: {
      initData,
    },
  }
}
```


issue: https://github.com/vercel/next.js/issues/13910

**Reply by AHTHneeuhl** [reply]
Is this working?

**Reply by mahmir217** [reply]
not working :(

**Reply by terryli0095** [reply]
Even if getServerSideProps is just an empty function, this problem persists. The 5+ seconds seem to suggest that it has to be a server warmup issue. Perhaps a cron job that pings the vercel server?

**Reply by smakosh** [reply]
Or switch the runtime to experimental edge

**Comment by andrelom** [comment]
I've noticed this as well

**Comment by msalmanshahid** [comment]
+1

**Comment by hamaike-biz** [comment]
I was having the same problem with Next.js on GAE.

I was doing Server side rendering of meta tags only for OGP.
I was using Mui's NoSsr to render the body text, and when I switched this to next/dynamic, the initial rendering, which used to take 27 seconds, was reduced to 200ms.

I don't know if this helps, but if you are using Mui, please give it a try.

**Reply by AlexHodo** [reply]
I used lazy load (next/dynamic) on the whole page content. Still having the same issue

**Comment by Alireza17224** [comment]
Same issue here

**Comment by ZacxDev** [comment]
+1

**Reply by ZacxDev** [reply]
This was happening on my login api route. I "fixed" it by sending a warmup request to the api route when the page loads so that when the actual login request is sent the route is already in memory.

**Reply by Shahriar-Shakil** [reply]
Could please explain the process

**Comment by fgrs** [comment]
same issue with nextjs 12.3.0 and firebase getDoc inside getServerSideProps, sometimes the page times out with:
`504: gateway timeout`
sometimes it doesn't and works straight away, totally random, does anybody know what's going on? I also installed Logalert hoping it  will spit out more details about the error but this is what I got:
`2022-12-01T21:48:59.237Z bfed084f-29af-4677-aa64-0532dacb75ef Task timed out after 10.02 seconds`

**Reply by AlexHodo** [reply]
Did you manage to fix it?

**Reply by fgrs** [reply]
nope and this is response from Vercel CS...

Hi there,

Thanks for reaching out to Vercel Support!

If you're getting a timeout from your serverless function, then this will be caused by something inside the request taking longer than expected. Nearly always this is caused by an upstream provider/database taking longer than expected or returning lots of data, causing your function to hit the timeout whilst processing it.

To fix this, we would recommend monitoring your upstream provider for latency, checking the queries for any unexpeted results and double checking that your serverless function location is located at near as possible to your upstream provider: https://vercel.com/docs/concepts/functions/serverless-functions/regions

Please let us know if you have any further questions or issues.

**Comment by xjustleoo** [comment]
+1

**Comment by terryli0095** [comment]
Even if getServerSideProps is just an empty function, this problem persists. The 5+ seconds seem to suggest that it has to be a server warmup issue. Perhaps a cron job that pings the vercel server?

**Comment by namanhvn** [comment]
It's 2023 now and do we have solution for this?

**Comment by mkosiedowski** [comment]
**TL;DR:** check the size of your bundled JS files generated for the path you are loading (`.next/server/pages/...`). If it's too big, the first load will be slow because Node needs to read that file to run the page's logic. That JS loading can take some time!

Simplest solution: use [dynamic imports](https://nextjs.org/docs/advanced-features/dynamic-import) **inside the `render` function** to make that page file smaller.

**Long version:**

I struggled with a similar issue in my app and spent a few hours tracing and debugging the internals of Next.js. In my case the first load of `/_next/data/.../page` endpoint took about 3 seconds when run on a remote server (docker image run inside a k8s cluster on GCP, but I am  not quite sure how relevant is that) and around 150ms when run locally on my Mac. Every subsequent request was way faster - around 150ms for a remote request (including network transfer) and around 30ms locally.

All that request did was checking a simple condition (login state) to either return a redirection to the login page or render the page's content. I couldn't understand why returning the redirection took up to 3 seconds! 

After hours of debugging I've found the cause. In my case the only piece of code that run significantly slower on the first run was the method `requirePage(...)` in `node_modules/next/dist/server/require.js`. Its whole purpose is to load the JS module responsible for the requested page. It does a simple `require(pagePath)` where the `pagePath` is the path to the file containing the code needed to render that page... or run only `getServerSideProps` and return the props as JSON (for the `_next/data` requests).

I've checked the file which was loaded for my path and it was... 140KB of JS (and of course some more `require` statements loading even more JS when that file was executed). Of course it contained not just the controller logic for that page, but also all of the components that were used if the page was to be rendered. All of that code was being read into Node's process memory even though at that point it wasn't needed. Loading that file wasn't that big of a thing when run on a modern Mac - with a fast SSD, fast RAM and lots of processing power. When run on a shared k8s cluster - it was very noticeable. Subsequent requests were much faster because [Node kept that loaded file in its memory cache and didn't really read the file again](https://nodejs.org/api/modules.html#caching).

So I've changed that page's main component and basically changed the code which looked like that:
```js
import PageContent from './PageContent';
export default () => {
  ... // do some stuff
  return <PageContent />;
}
```
with something like:
```js
export default () => {
  const PageContent = dynamic(() => import('./PageContent'));
  return <PageContent />
}
```

This way my bundled JS file with the page's controller turned out much smaller and the first load takes just few milliseconds more than subsequent loads. Case closed for me.

I hope this helps someone! 😄

**Reply by mvandergrift** [reply]
^^ this is perfect. This is exactly what was happening. I have some experience troubleshooting API routes on Vercel, but although the problem with getServerSideProps is similar the require of the render JS was confusing me. Thank you for taking the time to debug this!

**Reply by radoi90** [reply]
Doesn't this mean that the HTML generated in the SSR'ed page won't contain the contents of PageContent?

**Comment by Khalilsqu** [comment]
I am experiencing the same slowness when using getServerSideProps.

```
export default function Testing(props) {
  return <>{props.name}</>;
}

export async function getServerSideProps(context) {
  return {
    props: {
      name: "test",
    },
  };
}
```

Am I doing something wrong! this simple code takes about 7-10 seconds!

It is super fast in my local machine for both development and production and very laggy when deployed to Vercel!

**Comment by functor-dev** [comment]
After hours of research, I think the problem is caused by `*.nft.json` files.
From the docs:

> During next build, Next.js will use [@vercel/nft](https://github.com/vercel/nft) to statically analyze import, require, and fs usage to determine all files that a page might load.

So if you can reduce the size of `*.nft.json` files, then the page load is quicker.
I don't know how to fix it with the current version of Next.js, but my best way is using [Modularize Imports](https://nextjs.org/docs/advanced-features/compiler#modularize-imports).

My init page load time reduces from 5-7s to ~2s, I know it is not good enough but yeah, you're stuck with it for now.

I have an idea to load the `*.nft.json` file into node's cache but don't know how to do it, maybe you guys will come up with the solution.

**Comment by fgrs** [comment]
I think I partially solved this issue and after few days testing it seem to work more reliable but error still occurs, In my case when I looked at the page source it weighted around 30kb without the assets, I then moved almost every page component which don't need SSR into dynamic import leaving only h1 and main text in and this resulted in source size ~ 7.5kb There's still large google font style element inside the head section which I'm going to remove next and retest again.

Below is the error screenshot.

<img width="528" alt="Screenshot 2023-03-06 at 15 54 34" src="https://user-images.githubusercontent.com/3940061/223147637-e85c4de6-04e1-4925-bb49-8eaba8626c14.png">

UPDATE:
source code down to <4kb still erroring sometimes, I have no idea what else I can do :/

UPDATE2:
I just found out in Vercel/Functions/Function Region I had Washington selected which isn't close to my data source! updated and testing again...

**Reply by dakshAg** [reply]
Did it help?

**Reply by dejoma** [reply]
I had the same -- changed it and very little difference < 100ms so still insanely slow.

**Comment by censuradho** [comment]
i am facing the same error, any solution in 2023?

**Comment by khaleel737** [comment]
After numerious trials I finally think I figured it out, the problem is not with cold starting getserverside functions nor serverless api's, its actually in the imports, I've tested both default imports and named imports and it turns out named imports are the solution to getserversideprops as they only load the specific component when needed rather than the whole application with the component when you use the export default:

Solution:

Instead of using: export default Component or export default function Component use the named imports of export function Component, that will fix the issue of your speed and loading your page.

Good Luck everyone and I really hope this helps 😊

**Reply by dejoma** [reply]
I don't think that this is the solution since people are reporting examples without any imports at all, just exporting the page?

**Comment by EOEboh** [comment]
Seriously, the Vercel team need to bring a standard fix for this. It’s totally frustrating and makes me miss out on the advantage of SSR

**Comment by fgrs** [comment]
I did everything I can, dynamic imports, shrunk the code to less then 6kb, moved hosting from default Washington to Frankfurt as close to Firebase DB is possible, in fact it did improved UX slightly (less frequent errors) and I almost thought the issue was gone after few days testing it crashed again. Vercel tech support blames Firebase "Nearly always this is caused by an upstream provider/database taking longer than expected or returning lots of data" but I don't think this is the case, eventually I had to sadly give up SSR.

**Comment by namanhvn** [comment]
Too many comment and no response from Vercel team. I would like to have some updates

1. Why do we need to use `getServerSideProps `?
In my case, It's required when I use `https://github.com/i18next/next-i18next` package. However, I founded another solution to setup bilingual for my NextJS website using this instruction `https://viblo.asia/p/nextjs-da-ngon-ngu-khong-can-thu-vien-ngoai-Eb85oe0OZ2G`. Sorry to say that it written by Vietnamese.

2. Alternative for `getServerSideProps `?
Yes, I keep it simple in my page component like this

/pages/index.js
```React
import React from 'react';
import Home from 'components/Home';

const IndexPage = (): JSX.Element => {
  return <Home />;
};

export default IndexPage;
```

Then I fetch my data on page request within the component.
/components/Home/index.tsx

```React
import React, { useEffect } from 'react';

const Home = () => {

useEffect(() => {
// Fetch your data from server here.
},[])

return (
<></>
)

}

export default Home;
```

So in my case, I no longer need to use `getServerSideProps ` in any single page.

**Comment by Zain-ul-din** [comment]
![image](https://user-images.githubusercontent.com/78583049/230744716-d6d64394-0a78-44ee-b2e6-52d07252637b.png)
Same issue. But, No reply from the next js team so far. ☹

**Comment by doorman02** [comment]
I am using the Vercel Pro plan and I am running into this issue. I tried most of the suggestions using SSR and it seems it causes an inital load time delay of 5-7 seconds even if there is no logic inside getServerSideProps. However after the inital load its very fast on different devices... but only until the next cold start. Since this is an old issue, is there an explanation why this cold start occurs? Are there any future plans to solve this issue? Thanks!

**Reply by smakosh** [reply]
Switch to the edge runtime: https://nextjs.org/docs/advanced-features/react-18/switchable-runtime

**Comment by doorman02** [comment]
hi @smakosh thanks for the suggestion. Switching to the edge runtime sounds promising, however there are many packages in the solution that rely on the dom and the edge support for some node packages is be bit a limited at the moment and it's not that easy to migrate. Do you know if Vercel is looking into the cold start issue? Why is the cold start occuring, does the app not have an always running instance when using ssr? Thanks again!

**Comment by malimccalla** [comment]
I'm having the same issues. A page with `getServerSideProps` consistently takes 1-3 seconds to load. Switching  from `getServerSideProps` to `getInitialProps` fixed the issue for me

**Reply by dejoma** [reply]
However NextJS docs mention that getInitialProps is legacy:

> Note: getInitialProps is a legacy API. We recommend using [getStaticProps](https://nextjs.org/docs/pages/building-your-application/data-fetching/get-static-props) or [getServerSideProps](https://nextjs.org/docs/pages/building-your-application/data-fetching/get-server-side-props) instead.

**Comment by Alireza17224** [comment]
hi @malimccalla, Why use getInitialProps?
I mean how it solved the problem?
the getInitialProps will execute first on the server side but second on the client side which can make a kind of security issue if you are using some stuff to connect to your DB, ...

**Reply by malimccalla** [reply]
Am not quite sure how or why it solved the issue but it did. Fortunately I don't do any database connection stuff in getinitialprops

**Comment by denvudd** [comment]
Has anyone found a solution to this problem after all? I'm experiencing the same issue. I have a page that uses dynamic routing and fetches data through RTK Query in getServerSideProps (using next redux wrapper). It takes about 7 seconds to load the page if it's the first time. After the initial load, it loads faster (around 2 seconds). However, once the cache is cleared, it takes a long time to load again. I can't figure out what the problem is. In the DevTools Network tab, during the loading, it simply shows the JSON file in a pending state, nothing unusual. In my case, it's not specific to the development mode either. The page loads slowly on Vercel for the first time as well.

Can anyone at least provide a clue about the cause? :(

My package.json:
```
{
  "name": "react-tbmovies",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "@ant-design/icons": "^5.0.1",
    "@reduxjs/toolkit": "^1.9.5",
    "@types/node": "20.1.3",
    "@types/react": "18.2.6",
    "@types/react-dom": "18.2.4",
    "ahooks": "^3.7.7",
    "antd": "^5.4.7",
    "classnames": "^2.3.2",
    "clipboard": "^2.0.11",
    "dayjs": "^1.11.7",
    "lodash.debounce": "^4.0.8",
    "next": "13.4.2",
    "next-redux-wrapper": "^8.1.0",
    "numeral": "^2.0.6",
    "react": "18.2.0",
    "react-content-loader": "^6.2.1",
    "react-dom": "18.2.0",
    "react-redux": "^8.0.5",
    "react-youtube": "^10.1.0",
    "sass": "^1.62.1",
    "typescript": "5.0.4"
  },
  "devDependencies": {
    "@types/lodash.debounce": "^4.0.7",
    "@types/numeral": "^2.0.2"
  }
}
```

Page component [id].tsx:
```
import React from "react";
import DetailLayout from "@/layouts/DetailsLayout";
import MovieDetailsBlock from "@/components/MovieDetailsBlock/MovieDetailsBlock";
import {
  getMovieDetails,
  getRunningQueriesThunk,
} from "@/redux/api/movies/slice";
import { wrapper } from "@/redux/store";

interface MovieDetails {
  id: number;
}

export const getServerSideProps = wrapper.getServerSideProps(
  (store) => async (context) => {
    const { id } = context.query;

    if (typeof id === "string") {
      await store.dispatch(
        getMovieDetails.initiate({ id, params: "language=uk-UA&page=1" })
      );
    }

    await Promise.all(store.dispatch(getRunningQueriesThunk()));

    return {
      props: { id: id },
    };
  }
);

const MovideDetails: React.FC<MovieDetails> = ({ id }) => {
  const { data } = getMovieDetails.useQuery({
    id,
    params: "language=uk-UA&page=1",
  });

  return (
    <>
      <DetailLayout>
        {data && <MovieDetailsBlock id={id} data={data} />}
      </DetailLayout>
    </>
  );
};

export default MovideDetails;
```

DevTools Network during page load:
1. Pending state for JSON:
![image](https://github.com/vercel/next.js/assets/68691654/d3202270-3103-4e95-9928-6f36720fa8d3)

2. Fulfilled state for JSON and call information:
![image](https://github.com/vercel/next.js/assets/68691654/37f96b04-cb54-415f-8155-bbccc78d05ac)

**Reply by amitav13** [reply]
Is there a way to manually clear the cache to reproduce this?

**Reply by razgraf** [reply]
If you build and re-deploy the app it will count as a new bundle therefore a re-fetch should be triggered. Purging the cache directly from vercel (`settings/data-cache`) could help but not sure it forces an update on served content too.

**Comment by code-xhyun** [comment]
So no one has a solution?

**Reply by dejoma** [reply]
I removed getServerSideProps and now I use a custom hook with react query that calls an api endpoint 🥲

**Reply by denvudd** [reply]
However, this does not solve the problem because is no SSR :(

**Reply by dejoma** [reply]
I know 😢 but it's better than useEffect with an empty dependency array and definitely way more performant than the getServerSideProps..

**Comment by radoi90** [comment]
@cramforce it would be great to get some comms from the Vercel eng team on this. It's been a few years since this issue was raised and it would help to get official NextJS or Vercel team input on this. I understand the official plan is to encourage everyone to transition over to the app directory and server components, however it would still be helpful to understand what the core issue was here so that the community might look for a way to mitigate the problem.

**Comment by cifuwu** [comment]
I have a question, is this problem solved in the new app router? I am analyzing whether it is convenient to migrate the app router

**Comment by TaeyoonKwon** [comment]
Is this project abandonned?

**Reply by voidberg1** [reply]
probably yes

**Comment by Alireza17224** [comment]
Try the new App Router. It might work ...

**Comment by VinodhPIT** [comment]
Im using the latest Next 13 same issue

**Comment by AbdelrhmanX7** [comment]
This is all I have, and it takes around 1–2 seconds to load the page.

```javascript
export function getServerSideProps({ query }) {
  return {
    props: {
      query,
    },
  };
}
```

**Reply by MauroSerrano-dev** [reply]
Same issue here

**Comment by timfuhrmann** [comment]
In a lot of cases, this is most likely due to serverless function cold starts and the Node.js runtime booting. It depends on various factors, such as the dependencies inside your function. By switching to `egde` runtime you can improve the performance to near-zero cold starts. 

[This article](https://vercel.com/guides/how-can-i-improve-serverless-function-lambda-cold-start-performance-on-vercel) describes the issue for projects hosted on Vercel.

**Reply by Untel** [reply]
I don't think it's about cold start.
It was my first doubt, so i deployed in a dedicated server, and it was even slower

**Comment by mareksuscak** [comment]
We're currently experiencing this outside of Vercel, in standalone mode where Next is deployed in a Docker container running on Cloud Run (in CPU always allocated configuration with minimum number of instances set to 1 so cold starts should be minimal). There's got to be something else that's causing this - either the number of dependencies or `getInitialProps` in the custom Document? Could that be the culprit. We are pulling some data but navigating between pages takes 2-4s when clicking a Next Link whereas reloading the page takes a lot less (~1s). We tried the latest Canary release (13.4.20) to no avail. We use Pages Router, not the App Router.

**Reply by mareksuscak** [reply]
As a quick follow up for anyone who stumbled upon this comment, we managed to resolve this issue and it was a 3rd party library (mobx-react):
https://github.com/vercel/next.js/issues/53624#issuecomment-1713578575

**Comment by denvudd** [comment]
For those who are still suffering from this issue, I want to share my experience. My cold start time was about 7 seconds on a new project. 

Now I can say that:
- Using App Router in another project, I don't have this issue in DEV mode or production.
- Using `experimental-edge` in Pages Router helped reduce the cold start time from 7 seconds to 0.2 seconds (approximately, results may vary). 

In any case, when researching this issue, I found many similar issues that are still unresolved, and I couldn't find any comments from Vercel. I hope you all will find the solution to this problem. Good luck!

Relative:
[My Project's Issue](https://github.com/denvudd/react-dbmovies.github.io/issues/2)
[Slow initial page load with getServerSideProps](https://github.com/vercel/next.js/discussions/50783)
[Super slow page load times in development environment](https://github.com/vercel/next.js/discussions/17977) 

By the way, for some reason, I can't find the relevant documentation for the `experimental-edge` feature in Next.js. Here's a mention of it in the blog: https://nextjs.org/blog/next-12-2#edge-api-routes-experimental

**Reply by Jairwin-L** [reply]
But most people here are probably looking for solutions under page router, right? Not app router.

my scene is mainly for seo

**Comment by nicholasoxford** [comment]
Im having issues with `"next": "13.5.6",`

What version of node is everyone use? Were using Apollo query so not sure experimental-edge is any switch for us

**Comment by marharyta** [comment]
Same here. `"next": "13.4.19",`

**Comment by doorman02** [comment]
any updates on this? Vercel is a great service but for those not able to use Edge for the time being I wonder what options we have here. Since this only happens the first time, it must be a cold start issue right? Question is if its related to loading nodejs or just the app. For subscribing Vercel users it would be nice to get some feedback on what is really causing the delay even if it will not be fixed. Could this be solved with a different hosting plan? Is there an easy way, maybe a built in feature to keep the app warm? Thanks in advance for any feedback.

**Comment by TomasSestak** [comment]
pls any solution? this is unusable

**Comment by patched-admin** [comment]
We are facing this issue too. Looking at using the edge runtime but that would require us to rearchitect our app a bit. Hope a fix is prioritized.

**Comment by fgrs** [comment]
my solution was to leave vercel and move to amplify, no issues since.

**Comment by morgansson** [comment]
In my case, it was making my **/blog** page very slow, as I'm getting posts from an API, and I wanted to use serverSideProps for SEO, but when people already navigated the main page, or any other page, I get the posts using useEffect instead, and it was still taking ages to load.

I did a little, and bizarre hack to prevent this (which doesn't totally fix it, but it makes it unnoticeable).

`/pages/blog/preload.js`
```javascript

export default function Preload(){
  return null;
}
export const getServerSideProps = () => {
  return {
    props: {
      empty: true
    },
  };
};
```
and then anywhere in your layout:

```javascript
  useEffect(() => {
    fetch("/blog/preload");
  }, []);
  ```

**Comment by Matergi** [comment]
any solution? 😡

**Comment by malimccalla** [comment]
This is still the solve for me 

https://github.com/vercel/next.js/discussions/12447#discussioncomment-5897804

it uses the old API but solves the issue. Hoping this gets addressed

**Comment by liho00** [comment]
using next js 14+ same issues... planning to switch to remix js... 😭

**Comment by censuradho** [comment]
I had to migrate my project to the app folder, this only solved my problem 🤡

**Comment by nightwalker89** [comment]
without the cache it very slow always > 500ms. with cache it is under 50ms
```
export const getServerSideProps: GetServerSideProps<ResourcePageProps> = async ({ params, res, }) => {
    const resource = resources.find((r) => r.slug === params?.slug)

    if (!resource) {
        return { notFound: true }
    }

    res.setHeader(
        'Cache-Control',
        'public, s-maxage=10, stale-while-revalidate=59'
    )

    return { props: { resource, title: resource.name } }
}
```
